Add an option to getWriteableTensorData to avoid copy CUDA tensor to CPU

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44418 [WIP] Use streams from pool on RPC callees
* **#46524 Add an option to getWriteableTensorData to avoid copy CUDA tensor to CPU**
* #46523 Enable TP_USE_CUDA and TP_ENABLE_CUDA_IPC

Differential Revision: [D24392794](https://our.internmc.facebook.com/intern/diff/D24392794)

mrshenli committed:
Add an option to getWriteableTensorData to avoid copy CUDA tensor to CPU

[ghstack-poisoned]

lw reviewed (approved):
Looks good to me, but perhaps someone from the JIT team should also take a look?

mrshenli commented:
Hey @wanchaol, I added this because we need to access `WritableTensorData` in RPC and would like to avoid the cpu-to-gpu copy. Please let me know if this looks OK to you. If not, we can also try expose `WritableTensorData` struct from JIT or replicate the logic in RPC.

mrshenli committed:
Update on "Add an option to getWriteableTensorData to avoid copy CUDA tensor to CPU"




[ghstack-poisoned]

mrshenli committed:
Update on "Add an option to getWriteableTensorData to avoid copy CUDA tensor to CPU"


Differential Revision: [D24392794](https://our.internmc.facebook.com/intern/diff/D24392794)

[ghstack-poisoned]

wanchaol commented on torch/csrc/jit/serialization/pickler.cpp:
```diff
@@ -596,12 +596,14 @@ void Pickler::pushTuple(const IValue& ivalue) {
-WriteableTensorData getWriteableTensorData(const at::Tensor& tensor) {
+WriteableTensorData getWriteableTensorData(
+    const at::Tensor& tensor,
+    bool toCpu) {
```
nit: to_cpu?

wanchaol reviewed (approved):
This looks good to me.

mrshenli committed:
Update on "Add an option to getWriteableTensorData to avoid copy CUDA tensor to CPU"


Differential Revision: [D24392794](https://our.internmc.facebook.com/intern/diff/D24392794)

[ghstack-poisoned]

mrshenli committed:
Update on "Add an option to getWriteableTensorData to avoid copy CUDA tensor to CPU"


Differential Revision: [D24392794](https://our.internmc.facebook.com/intern/diff/D24392794)

[ghstack-poisoned]
